deps/jerry/jerry-ext/handle-scope/handle-scope.c

```diff
@@ -0,0 +1,284 @@
+jerryx_handle_scope_status
+jerryx_close_handle_scope (jerryx_handle_scope scope)
+{
+  /**
+   * Release all handles related to given scope and its child scopes
+   */
+  jerryx_handle_scope a_scope = scope;
+  do
+  {
+    jerryx_handle_scope_release_handles (a_scope);
+    jerryx_handle_scope child = jerryx_handle_scope_get_child (a_scope);
+    jerryx_handle_scope_free (a_scope);
+    a_scope = child;
+  } while (a_scope != NULL);
+
+  return jerryx_handle_scope_ok;
+}
+
+
+/**
+ * Opens a new handle scope from which one object can be promoted to the outer scope
+ * and attach it to current global scope as a child scope.
+ *
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_open_escapable_handle_scope (jerryx_handle_scope *result)
+{
+  return jerryx_open_handle_scope (result);
+}
+
+
+/**
+ * Close the scope and its child scopes and release all jerry values that
+ * resides in the scopes.
+ * Scopes must be closed in the reverse order from which they were created.
+ *
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_close_escapable_handle_scope (jerryx_handle_scope scope)
+{
+  return jerryx_close_handle_scope (scope);
+}
+
+
+/**
+ * Escape a jerry value from the scope to its parent scope.
+ * An assertion of if parent exists shall be made before invoking this function.
+ *
+ * @returns escaped jerry value id
+ */
+jerry_value_t
+jerryx_hand_scope_escape_handle_from_prelist (jerryx_handle_scope scope, size_t idx)
+{
+  jerry_value_t jval = scope->handle_prelist[idx];
+  jerryx_create_handle_in_scope (scope->handle_prelist[idx], jerryx_handle_scope_get_parent (scope));
+
+  if (scope->handle_count > JERRY_X_HANDLE_SCOPE_PRELIST_HANDLE_COUNT)
+  {
+    jerryx_handle_t *handle = scope->handle_ptr;
+    scope->handle_ptr = handle->sibling;
+    scope->handle_prelist[idx] = handle->jval;
+    scope->handle_count -= 1;
+    return jval;
+  }
+
+  if (idx < JERRY_X_HANDLE_SCOPE_PRELIST_HANDLE_COUNT - 1)
+  {
+    scope->handle_prelist[idx] = scope->handle_prelist[scope->handle_count - 1];
+  }
+  scope->handle_count -= 1;
+  return jval;
+}
+
+
+/**
+ * Promotes the handle to the JavaScript object so that it is valid for the lifetime of
+ * the outer scope. It can only be called once per scope. If it is called more than
+ * once an error will be returned.
+ *
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_escape_handle (jerryx_escapable_handle_scope scope,
+                      jerry_value_t escapee,
+                      jerry_value_t *result)
+{
+  jerryx_handle_scope parent = jerryx_handle_scope_get_parent (scope);
+  if (parent == NULL)
+  {
+    return jerryx_handle_scope_mismatch;
+  }
+
+  bool found = false;
+  do
+  {
+    size_t found_idx = 0;
+    for (size_t idx = 0; idx < JERRY_X_HANDLE_SCOPE_PRELIST_HANDLE_COUNT; idx++)
+    {
+      if (escapee == scope->handle_prelist[idx])
```

Comment:
How about checking this in reverse order? The `handle_prelist` is a LIFO(stack), searching from tail might be always faster than head, right?

Reply:
There is no actually out order of handle list, since we don't have a mechanism for remove a handle from a scope. The escaping doesn't guarantee if the escapee is the last one that added to the scope either.
Or is there any potential that an escapee is more possibly added to the scope lastly?